[이지예] 3주차 로또 게임 구현 완료

jiyeeeah opened the pull request:
## 🎱 로또 게임 기능 구현 목록

### 📌 로또 구입

- [x] 로또 구입 금액 입력 받기
  - [x] 1,000원 단위로 입력 받아야 함. 1,000원으로 나누어 떨어지지 않으면 예외 처리 & 다시 입력 받기
- [x] 구입 금액으로 발행된 로또의 개수(로또 한 장당 1,000원)출력
- [x] 발행된 로또 번호 출력 (1 ~ 45, 오름차순, 중복 x)

### 📌 로또 당첨 번호 입력

- [x] 당첨 번호 입력 받기
  - [x] 잘못된 값 입력하면 throw문으로 예외 발생 & 다시 입력 받기
    - 숫자여야 함
    - 6개의 번호여야 함
    - 모든 번호가 1~45 사이여야 함
    - 숫자가 중복되면 안됨
- [x] 보너스 번호 입력 받기
  - [x] 잘못된 값 입력하면 throw문으로 예외 발생 & 다시 입력 받기
    - 숫자여야 함
    - 1~45 사이여야 함
    - 이전에 입력한 로또 번호들과 중복되면 안됨

### 📌 로또 결과 출력

- [x] 당첨 통계 출력 (각 액수마다 몇장씩 당첨됐는지)
- [x] 수익률 출력 (소수점 둘째 자리에서 반올림)

### 📌 프로그래밍 요구 사항

- [x] 함수 또는 메서드의 길이가 15라인을 넘어가지 않도록 (한 가지 일만 잘 하도록)
- [x] else 지양 (쓰는 게 더 좋은 경우는 스스로 판단)
- [x] 단위 테스트 구현 (UI 로직 제외)
  - 핵심 로직을 구현하는 코드와 UI를 담당하는 로직 구분

---
## 🎱 코드 구성 설명

이번 주차에는 주석으로 최대한 설명해봤습니다! 참고해주시면 감사하겠습니다 :)
(절대 시간 부족해서.. 그런 것 ~~맞습니다~~...ㅠㅠ)

---
### 🙌🏻 리뷰어분들께 

1. 객체지향... 1주차보다 더 깊게 다뤄본 것 같아요! 이번에는 MVC 패턴을 활용해보려고 했습니다. 이 패턴이 잘 드러났는지, 어디선가 어긋나는 부분은 없는지 궁금합니다
2. 단위 테스트가 뭔지 아직도 잘 모르겠지만 최대한 구현해보려고 했어요! 이렇게 하는게 맞는 건지 모르겠지만... 관련해서 피드백 자유롭게 주시면 감사하겠습니다ㅠㅠ
3. 이번에도 commit에 신경썼는데 잘 되었는지 모르겠네요.. 이 부분도 신경써서 봐주시면 매우 감사하겠습니다!

항상 꼼꼼한 리뷰 감사드립니다 ☺️

jiyeeeah committed:
[docs] 기능 구현 목록 작성

jiyeeeah committed:
[feat] MVC 패턴으로 파일 구성

jiyeeeah committed:
[feat] 로또 구입 금액 입력 받기

- 로또 구입 금액을 input(View)의 inputMoney 메서드를 활용해 받고,
  ticket(Model)에서 유효성 검사를 거치도록 함
- Constroller에 입력값 받고, ticket 데이터 만드는 로직이 담겨있음

jiyeeeah committed:
[feat] 구매한 로또 개수 출력

- 구입금액으로 발행된 로또의 개수를 출력하는 기능 구현

- Ticket(Model)에 데이터가 만들어질 때, 내부 메서드(calculateAmount)를 통해
  티켓 개수(amount)를 계산하도록 함.
  - Ticket(Model)의 amount라는 필드는 getAmount라는 메서드를 통해 접근할 수 있도록 함

- Output(View)에 n개를 구매했다는 문구를 출력하는 메서드 추가

- Controller에서 Ticket의 amount를 getAmount()를 통해 가져오고
  Output의 outputTicketAmount를 통해 출력하도록 함

jiyeeeah committed:
[feat] 발행된 로또 번호 출력

- Ticket(Model)안에 랜덤한 로또 번호를 넣어주는 메서드(getOneTicket) 구현
- getOneTicket를 amount만큼 반복하여 tickets라는 필드에 넣어줌
- tickets라는 필드는 private하기 때문에 getTickets()라는 메서드를 통해서만 접근 가능

- Output(View)안에 로또를 출력하는 기능 구현

- Controller에서 ticket출력할 수 있도록 함

jiyeeeah committed:
[feat] 당첨 번호 입력 받기

Model
- Lotto 클래스에 당첨 번호가 담기도록 함
- 여기서 유효성 검사도 함
  - #validate에는 중복을 찾는 로직 추가
  - #validateEachNum에서는 각 숫자가 올바른지 확인

View
- Input에 inputWinningNums메서드를 통해 입력값 받음
- inputWinningNums에서 받을 때, 쉼표를 기준으로 나눠지도록 함

Controller
- inputWinningNum이라는 새로운 메서드를 통해 당첨 번호 입력 받고
  Lotto(Model)에 저장할 수 있도록 함

jiyeeeah committed:
[feat] 보너스 번호 입력 받기

Model
- BonusNumber 클래스를 통해 보너스 번호 데이터를 저장
  - 이 안에서 validation 검사 수행
    - validation 검사를 위해 winningNumbers를 전달받음(controller에서)
  - bonusNumber 필드는 private이기 때문에
    getBonusNumber 메서드를 통해서만 접근할 수 있음

View
- 보너스 번호를 입력받는 input 인터페이스 구현

Controller
- 보너스 번호를 받기 위해 인터페이스 실행
- 입력받은 보너스 번호와 이전에 입력받았던 winningNums를 함께 BonusNumber 모델에 넘겨줌

jiyeeeah committed:
[feat] 당첨 통계 출력

- Controller에서 Model의 값들을 활용하여 당첨 계산 기능 구현

jiyeeeah committed:
[refactor] 게임 룰 관련 상수화

- 게임 룰에 사용되는 숫자들 상수화
- 최종 등수도 상수화

jiyeeeah committed:
[refactor] 게임 상 메세지들 상수화

- 게임에서 쓰이는 입출력 메세지, 에러 메세지 모두 상수화

jiyeeeah committed:
[feat] 수익률 출력

- 당첨 통계에 따른 수익률 출력
- 수익률 계산을 위해 필요한 purchaseMoney를 필드로 선언하고
  userRank를 활용하여 수익을 계산
- 최종 결과를 출력하는 printResult메서드에 수익률도 출력하도록 추가

jiyeeeah committed:
[refactor] 티켓 배열 문자열로 출력되도록 수정

- 티켓 배열이 그대로 console에 출력되면서 jest오류가 생김
- 배열을 문자열의 형태로 변환해서 출력

jiyeeeah committed:
[feat] LottoTest 추가

- Lotto 클래스에 대한 유효성 검사 test 코드 추가

jiyeeeah committed:
[feat] TicketTest 추가

- 구매 금액을 입력시 구입 금액에 관한 유효성 검사와
  금액에 따른 티켓 개수의 계산을 검증하기 위한 test 코드 작성

jiyeeeah committed:
[feat] BonusNumberTest 추가

- BonusNumber 클래스를 검증하는 test코드 추가

jiyeeeah committed:
[refactor] 당첨 통계를 관리하는 Model(WinTicket) 분리

- 기존에 Controller에 존재하던 당첨 통계를 계산하고,
  관련된 데이터를 관리하는 로직을 분리함
- 이로써 Controller에는 View와 Model을 다루는 로직만 남아있게 됨

jiyeeeah committed:
[docs] 주석 추가

- constants, Model, View, Controller에 JSDoc 형태의 주석을 추가

jiyeeeah committed:
[refactor] 불필요한 필드 제거

- purchaseMoney는 WinTicket내에서 tickets의 length로 충분히 구할 수 있는 값이므로
  constructor에 파라미터로 보내주지 않아도 된다.
- 따라서, Controller에서 여러번 쓰일 것을 고려해 필드로 선언했던 purchaseMoney를 제거함

jiyeeeah committed:
[feat] WinTicketTest 추가

- 새로 추가한 WinTicket에 대한 Test를 추가함
- 정해진 로또 번호들과 당첨 번호, 보너스번호를 주어줬을 때 계산이 올바른지 확인

jiyeeeah committed:
[docs] 프로그래밍 요구 사항 체크표시

wo-o29 commented on src/constants/rules.js:
```diff
@@ -0,0 +1,17 @@
+/** 로또 게임의 룰과 관련된 상수들입니다. */
+export const LOTTO = Object.freeze({
```
```suggestion
export const LOTTO_RULES = Object.freeze({
```
`LOTTO`라는 변수명으로 보고 상수화 되어있는 객체라고 파악하기는 어려울 것 같아서 `LOTTO_RULES`와 같이 수정해보는건 어떠신가요?!

wo-o29 commented on src/constants/rules.js:
```diff
@@ -0,0 +1,17 @@
+/** 로또 게임의 룰과 관련된 상수들입니다. */
+export const LOTTO = Object.freeze({
+  NUM_COUNT: 6,
```
```suggestion
 LENGTH: 6,
```
`LOTTO_RULES.NUM_COUNT`가 로또 번호 길이를 의미하는 것 같은데 조금 더 명확하게 `LOTTO_RULES.LENGTH` 이렇게 사용해보는건 어떨까요??

wo-o29 commented on src/View/Output.js:
```diff
@@ -0,0 +1,32 @@
+    for (let rank = 5; rank > 0; rank -= 1) {
+      const COUNT = resultArr[rank];
+      const RANK_STRING = RANK[rank].STRING;
+      MissionUtils.Console.print(RANK_STRING + " - " + COUNT + "개");
```
지예님은 템플릿 리터럴이랑 이렇게 `+`연산자를 사용하는 기준이 있으신가요?

jiyeeeah replied:
아 이거 고친다는걸 까먹었네요... ㅋㅋㅋㅋ ㅠㅠ 저는 리터럴을 선호하긴 합니다만 취향차이 아닐까여? ㅎㅎ

wo-o29 replied:
아하 그렇군요!

저는 개인적으로 깔끔하게 `+`로 해결할 수 있으면 `+`를 사용하고, 중간에 다른 문자가 들어가거나하는 경우에는 리터럴 문법을 사용하자라고 기준을 세웠던 것 같아요🙃

```js
const a = "안녕";
const b = "하세요";

const hello = a + b; // + 연산자로 해결
const helloDash = `${a}-${b}`; // 중간에 다른 문자가 포함되어 리터럴 문법 사용
```

jiyeeeah replied:
마자요! 저도 그렇게 보통 해서 `RANK_STRING + " - " + COUNT + "개"`이 경우에는 중간중간 문자가 포함되어서 리터럴로 하려고 했는데 어쩌다가 `+`로 해버린걸 수정을 안했네요 ㅎㅎ

wo-o29 commented on src/View/Input.js:
```diff
@@ -0,0 +1,40 @@
+    try {
+      const input = await MissionUtils.Console.readLineAsync(INPUT_MESSAGES.BONUS_NUM);
+      return input;
+    } catch (error) {
```
여기 메서드에서 `catch`에 예외처리가 되려면 어떻게 해야하나요??

제 생각에는 여기로 들어올 수는 없을 것 같은데 맞나요??

jiyeeeah replied:
질문을 이해를 못했어요...ㅠㅠ `catch`문 안에서 `throw`문으로 예외 발생시켰을 때를 말씀하시는건가요?

wo-o29 replied:
지금 인풋을 받는 메서드에서 `catch`문 안에서 예외처리가 되려면 어떻게 해야할까요??

제 생각에는 지금 이 코드에서는 `catch`문 안으로 들어올 수 없다고 생각이 들더라구요!_!

jiyeeeah replied:
아 여기에 `try catch`문을 사용한건 혹시 `readLineAsync`에서 에러가 생겼을 때를 대비하고 싶어서 작성했어요! 정상적인 상황이라면 `catch`문 안으로 못 들어오는게 맞습니당

wo-o29 commented on src/View/Input.js:
```diff
@@ -0,0 +1,40 @@
+  /** 로또 총 구매 금액을 입력받습니다. */
+  async inputMoney() {
+    try {
+      const input = await MissionUtils.Console.readLineAsync(INPUT_MESSAGES.PURCHASE_MONEY);
+      return input;
```
```suggestion
      return MissionUtils.Console.readLineAsync(INPUT_MESSAGES.PURCHASE_MONEY);
```
이렇게 사용하는거랑 기존 코드랑은 어떤 차이가 있을까요??

jiyeeeah replied:
오우 차이가 없네요 ㅎㅎ 바꿔도 괜찮을 것 같아용 :)

wo-o29 commented on src/Model/WinTicket.js:
```diff
@@ -0,0 +1,60 @@
+  /** 당첨 통계를 계산합니다. */
+  #calculateMatch(ticket, winningNums, bonusNum) {
+    const matchNums = ticket.filter((num) => winningNums.includes(num));
+    if (matchNums.length < 3) return;
+    else if (matchNums.length === 6) return 1;
+    else if (matchNums.length === 5 && ticket.includes(bonusNum)) return 2;
+    else return 8 - matchNums.length;
```
여기 부분을 `early return`을 사용하면 else를 지양할 수 있을 것 같은데 어떻게 생각하시나요??

`8 - matchNums.length`는 어떤 의미인가요??

jiyeeeah replied:
사실 다 else를 제거해도 되는데 그냥 else를 확실하게 보여주고 싶어서 넣었어요 ㅎㅎ
`8 - matchNums.length` 이거는 3개가 match되었을 때는 5등, 4개가 맞으면 4등 ... 이걸 표현하고 싶었습니다

wo-o29 replied:
오호 그렇군요 `8 - matchNums.length` 이 부분은 다른 사람이 봤을 때 어떤 의도인지 파악하기 어려울 것 같다고 생각이 들어서 따로 객체를 만들거나 메서드로 분리해봐도 좋을 것 같아요!
```js
const getRank = (matchCount) => {
   // 내부에서 if문 혹은 switch문으로 분기
}

const rank = {
   4: 4
   3: 5
}
const userRank = rank[matchCount];
```

jiyeeeah replied:
저두 `8 - matchNums.length` 이 부분 작성하면서 알아보기 어려울거라고 생각하긴 했어요.. 따로 메서드를 분리해보는 것 괜찮은 방법인 듯 합니다! 고민해볼게용 감사합니다 👍

wo-o29 commented on src/Model/WinTicket.js:
```diff
@@ -0,0 +1,60 @@
+   * @param {number} purchaseMoney
+   */
+  constructor(tickets, winningNums, bonusNum) {
+    for (let ticket of tickets) {
```
```suggestion
    for (const ticket of tickets) {
```
`let`을 사용하신 이유가 있으실까요??

jiyeeeah replied:
흠.. 습관인 것 같군요ㅠㅠ 감사합니다..!

teagu123 replied:
저도 잘 몰라서 그러는데 여기에 let과 const의 차이가 있을까요??

jiyeeeah replied:
여기서 `ticket`을 `let`으로 선언하면 이후에 `ticket`을 수정할 수 있게 됩니당. 하지만 `for`문 안에서 `ticket`을 수정할 일도 없고 (어차피 배열이라서 수정이 되긴 하겠지만) 안정성을 위해 `const`를 사용하는게 좋을 것 같습니다!

wo-o29 commented on src/Model/Lotto.js:
```diff
@@ -0,0 +1,46 @@
+   */
+  constructor(numbers) {
+    this.#validate(numbers);
+    for (let number of numbers) {
```
```suggestion
    for (const number of numbers) {
```
여기도 `let` 키워드를 사용하신 이유가 있으신가요??

wo-o29 commented on __tests__/BonusNumberTest.js:
```diff
@@ -0,0 +1,34 @@
+    }).toThrow(ERROR_MESSAGES.ONLY_NUMBER);
+  });
+
+  const BONUS_DUPLICATE_CASE = [["1", ["1", "2", "3", "4", "5", "6"]]];
+  test.each(BONUS_DUPLICATE_CASE)(
+    "보너스 번호는 기존의 번호와 중복되면 안된다.",
+    (input, winningNums) => {
+      expect(() => {
+        new BonusNumber(input, winningNums);
+      }).toThrow(ERROR_MESSAGES.DUPLICATE);
+    }
+  );
```
![스크린샷 2024-10-03 오후 11 17 46](https://github.com/user-attachments/assets/a4f4b773-af97-4f48-aa96-5d9949ccc149)
지예님 코드를 실행했을 때 보너스 번호가 중복이 되는 것 같은데 테스트 코드는 통과하는 것 같네요...?!

jiyeeeah replied:
띠용..? 다시 보겠습니다..ㅠ

wo-o29 commented on __tests__/BonusNumberTest.js:
```diff
@@ -0,0 +1,34 @@
+  });
+
+  const BONUS_DUPLICATE_CASE = [["1", ["1", "2", "3", "4", "5", "6"]]];
+  test.each(BONUS_DUPLICATE_CASE)(
```
여기 테스트가 하나 밖에 없는 것 같은데 `each`를 사용하신 이유가 있으신가요??

jiyeeeah replied:
더 추가할 계획...이었다고 하면 믿어주실 수 있나요? ㅋㅋㅋ 추가해보겠습니당

wo-o29 replied:
열린 코드였군요ㅋㅋㅋ 제가 못알아봤습니다🙃

wo-o29 commented on __tests__/TicketTest.js:
```diff
@@ -0,0 +1,26 @@
+  ];
+  test.each(AMOUNT_TEST_CASE)("구입 금액에 따른 티켓 개수 계산 test", (input, amount) => {
+    const ticket = new Ticket(input);
+    expect(ticket.getAmount()).toEqual(amount);
```
```suggestion
    expect(ticket.getAmount()).toBe(amount);
```
원시 값을 비교하는 것 같은데 `toBe`가 아닌 `toEqual`을 사용하신 이유가 있으신가요??

jiyeeeah replied:
제가 Jest는 잘 몰라서.. 혹시 다른 점이 있나요..? 그냥 어디선가 봤어서 사용한 것 같아요 ㅎㅎ

wo-o29 replied:
`toBe`: `object.is` 사용, 원시 값 비교에 적합
`toEuqal`: 모든 속성을 재귀적으로 비교(깊은 비교), 객체나 배열 비교에 적합
저는 이렇게 생각했던 것 같아요! 자세한건 공식문서 참고해보셔도 좋을 것 같아요🙃
https://jestjs.io/docs/expect#tobevalue
https://jestjs.io/docs/expect#toequalvalue

jiyeeeah replied:
와우 감사합니다! 참고해볼게요! 👍

wo-o29 commented on __tests__/WinTicketTest.js:
```diff
@@ -0,0 +1,18 @@
+      42
+    );
+    expect(winTicket.getUserRank()).toEqual([0, 0, 1, 0, 0, 0]);
+    expect(winTicket.getProfitRate()).toEqual(750000);
```
여기 수익률을 구할 때 구매 금액이 없는 것 같은데 어떻게 수익률을 구할 수 있는건가요?? 
저의 좁은 시야에서는 여러번 봐도 잘 모르겠네요...😭

jiyeeeah replied:
`WinTicket` 클래스에서 보시면 `constructor`에서 `tickets`를 받는데 그거 `길이 * 1000` 해서 구매 금액을 내부에서 따로 구했습니다 ㅋㅋㅋ 다시 생각해보니 약간 야매네요.. `구매 금액`이랑 `구매 티켓 개수 * 단위`는 다를 수도 있는건데 🤔

wo-o29 replied:
아하 그렇군요 그 방법은 생각도 못했던 것 같네요🤔

wo-o29 commented on __tests__/LottoTest.js:
```diff
@@ -15,4 +17,24 @@ describe("로또 클래스 테스트", () => {
 
   // 아래에 추가 테스트 작성 가능
+  test("로또 번호에 숫자가 아닌 입력값이 있으면 예외가 발생한다.", () => {
+    expect(() => {
+      new Lotto(["1", "2", "3", "4", "5", "hi"]);
+    }).toThrow(ERROR_MESSAGES.ONLY_NUMBER);
```
```suggestion
    expect(() => new Lotto(["1", "2", "3", "4", "5", "hi"])).toThrow(ERROR_MESSAGES.ONLY_NUMBER);
```
여기서 중괄호를 사용하신 이유가 있으신가요??
if문에서는 중괄호를 생략하셔서 그걸 선호하시는 줄 알았는데 테스트 코드에서는 작성하신 것 같아서 그 이유가 궁금합니다!_!

jiyeeeah replied:
예시로 있었던 LottoTest.js 코드에 이렇게 되어있길래 복붙했습니다 :) 비슷해보이도록 하고 싶어서요 ㅋㅋㅋ

wo-o29 replied:
테스트 코드짜다보면 위에 작성했던거랑 조금만 바뀌면 되서 복붙 많이 하게 되더라구요🙃

wo-o29 commented on src/Controller/Controller.js:
```diff
@@ -0,0 +1,52 @@
+      this.bonusNumber = new BonusNumber(bonusNum, this.lotto.getWinningNumbers());
+    } catch (error) {
+      MissionUtils.Console.print(error.message);
+      await this.inputWinningNum();
```
지금 방식은 보너스 번호에서 틀려도 당첨 번호쪽부터 다시 받을 것 같아요!_!

jiyeeeah replied:
마자요! 일부러 그렇게 하긴 했는데.. 나누는게 좋을까요? 🤔 나누는게 좋겠군요... 감사합니다 ㅎㅎ

wo-o29 reviewed (approved):
고생 많으셨습니다👍⭐️💡🎉

수익률을 출력할 때 세자리마다 콤마(,)를 찍어도 좋을 것 같고, `number`, `num`을 둘 다 사용하시는 것 같은데 하나로 통일하면 코드의 일관성을 높일 수 있을 것 같아요!!

저도 리뷰하면서 생각났는데 구매 금액에 0원을 제한해봐도 좋을 것 같다고 생각이 드는 것 같네요🤔
그리고 에러 발생하면 해당 부분부터 입력을 받아야 하는 요구사항을 놓쳤던 것 같아서 리팩토링때 고쳐봐야겠네요…
지예님꺼 리뷰하면서 제 코드를 많이 되돌아보게 된 것 같네요 감사합니다!!

이번 리뷰에서는 저도 클래스 문법에 무지해서 조금 지예님 의견을 여쭤보는 리뷰가 많았는데 답변하시기 조금 내용이 길다고 생각하시는 부분은 이번 회의때 답변 주셔도 좋습니다🙃

그리고 지예님이 작성하신 코드 방식과 함수형 방식에 어떤 장단점이 있다고 느끼셨는지 공유해 주시면 좋을 것 같아요~!

wo-o29 commented:
> 2. 단위 테스트가 뭔지 아직도 잘 모르겠지만 최대한 구현해보려고 했어요! 이렇게 하는게 맞는 건지 모르겠지만... 관련해서 피드백 자유롭게 주시면 감사하겠습니다ㅠㅠ

저는 단위(Unit) 테스트를 가장 작은 단위, 하나의 함수나 메서드를 테스트한다고 생각 했던 것 같아요!
그리고 이 단위들을 여러개로 묶으면 통합 테스트, 일련의 프로세스 단위(로그인 단계, 회원가입 단계 등), 유저 플로우를 E2E 테스트라고 정의해봤던 것 같아요(제 개인적인 의견입니다...)

teagu123 commented on src/View/Input.js:
```diff
@@ -0,0 +1,40 @@
+import { INPUT_MESSAGES } from "../constants/message.js";
+
+/** 입력과 관련된 로직들을 관리하는 class입니다. */
+class Input {
+  /** 로또 총 구매 금액을 입력받습니다. */
+  async inputMoney() {
+    try {
+      const input = await MissionUtils.Console.readLineAsync(INPUT_MESSAGES.PURCHASE_MONEY);
+      return input;
+    } catch (error) {
+      MissionUtils.Console.print(error.message);
+      this.inputMoney();
+    }
+  }
+
+  /** 로또 당첨 번호를 입력받습니다. */
+  async inputWinningNums() {
+    try {
+      const input = await MissionUtils.Console.readLineAsync(INPUT_MESSAGES.WINNING_NUM);
+      return input.split(",");
+    } catch (error) {
+      MissionUtils.Console.print(error.message);
+      this.inputWinningNums();
+    }
+  }
+
+  /** 로또 보너스 번호를 입력받습니다. */
+  async inputBonusNum() {
+    try {
+      const input = await MissionUtils.Console.readLineAsync(INPUT_MESSAGES.BONUS_NUM);
+      return input;
+    } catch (error) {
+      MissionUtils.Console.print(error.message);
+      this.inputWinningNums();
+    }
+  }
+}
+
+export default Input;
```
이렇게 View를 분리하는거는 생각하지 못했는데 너무 재밌네요!!!

jiyeeeah replied:
흥미로운 접근법이긴 하죠..! 한번 이렇게 구현해보니까 뭔가 이런 작은 규모에서는 굳이 싶긴 하지만, 작은 규모에서 요런 구조 연습해 볼 수 있어서 좋았네용 ㅎㅎ

teagu123 commented on src/Model/Ticket.js:
```diff
@@ -0,0 +1,62 @@
+  }
+
+  /** 한개의 로또 티켓을 구매했을 때의 로또 번호를 구합니다. */
+  #buyOneTicket() {
+    const nums = MissionUtils.Random.pickUniqueNumbersInRange(
+      LOTTO.MIN_NUM,
+      LOTTO.MAX_NUM,
+      LOTTO.NUM_COUNT
+    );
+    nums.sort((a, b) => a - b);
+    this.#tickets.push(nums);
+  }
```
buyOneTicket을 따로 분리하신 이유가 있으실까요? 
개인적인 제 스타일에서 가독성을 보았을때 반복문을 이 클래스에 넣고 로또 번호 배열을 출력하는 클래스로 넣어도 하나의 함수안에 하나의 동작만 하고있다고 생각을해서 질문드립니다!! 작은 단위로 나누실려고 하신거같아서 그게 맞다면 너무 깔끔하네요

jiyeeeah replied:
'함수나 메서드가 하나의 기능만 하도록'을 실천해보고 싶었습니다! 근데 말씀하신대로 이 안에서 반복문을 돌려도 '하나의 기능'만 한다고 얘기할 수도 있을 것 같긴 하네용 ㅎㅎ 그냥 작은 단위로 하고 싶었나... 저도 이유를 잘 모르겠네요ㅠ 좀 더 고민해봐야겠습니다..! 감사합니다 ☺️

teagu123 commented on src/View/Output.js:
```diff
@@ -0,0 +1,32 @@
+  /** 당첨 통계와 수익률을 출력합니다. */
+  printResult(resultArr, profitRate) {
+    MissionUtils.Console.print(OUTPUT_MESSAGES.RESULT_STATEMENT);
+    for (let rank = 5; rank > 0; rank -= 1) {
```
```suggestion
    for (let rank = 5; rank > 0; rank --) {
```
코드를 보면서 다른곳은 for문에서 ++이런식으로 하시는데 여기는 -=이여서 혹시 --와 같은 동작을 한다면 
--로 통일은 어떨까요? 코드를 봤을때 대략 --가 될거같은데 제가 저렇게는 써보지않아서 조심스럽게 말씀드립니다 ㅎ

jiyeeeah replied:
엇 그렇네요! 통일성에 도움이 많이 될 것 같습니당! 감사합니다 ☺️ 👍

teagu123 reviewed (approved):
커밋을 봤을때 커밋을 통한 추적과 이 분이 어떤 식으로 구상했는지 알수있어서 놀랐습니다!
커밋을 이렇게 하는게 진짜 보기좋구나 이 생각을 많이했던거같아요 ☺️

전체적으로 개발을 너무 잘하신거같아서 많이 보고 배웠습니다. 
그리고 클래스형에서 프라이빗에 대한 부분을 잘 정의하고 클래스를 구상하시는거같아서 좋았습니다. 
고생하셨어요

jiyeeeah committed:
[refactor] return 문 축약

jiyeeeah committed:
[refactor] LOTTO -> LOTTO_RULES 네이밍 변경

- LOTTO라는 상수의 네이밍을 LOTTO_RULES로 변경해,
  게임의 규칙과 관련된 상수들을 객체로 담고 있음을 명시함

jiyeeeah committed:
[refactor] 표현법 통일

- 위에서는 i++로 표현하고 아래에선 rank -= 1로 표현함
  -> 이를 ++과 --로 통일시킴

jiyeeeah committed:
[refactor] for of문에서 let 대신 const 선언하도록

jiyeeeah committed:
[refactor] 당첨 번호와 보너스 번호 입력받는 메서드 분리

- controller에서 당첨 번호를 입력받는 메서드와
  보너스 번호를 입력받는 메서드를 분리함
- 보너스 번호를 입력 받을 때 예외가 발생해도 보너스번호만 다시 입력받게 됨

jiyeeeah committed:
[bugfix] BonusNumber 내 중복 예외처리의 조건 수정

- number는 string이고, winningNums는 number의 배열인데
  number로 된 배열 에서 string이 포함되어있는지 판단하는
  조건문은 항상 false를 return함
- 따라서, 조건문에 string형태인 number 파라미터를 Number로
  형변환하여 포함하는지 판단하도록 수정함

ssilver01 commented on src/constants/rules.js:
```diff
@@ -0,0 +1,17 @@
+
+/** 로또 게임의 당첨 등수별 출력 메시지와 가격을 담은 객체입니다. */
+export const RANK = Object.freeze({
+  1: { STRING: "6개 일치 (2,000,000,000원)", PRICE: 2000000000 },
+  2: { STRING: "5개 일치, 보너스 볼 일치 (30,000,000원)", PRICE: 30000000 },
+  3: { STRING: "5개 일치 (1,500,000원)", PRICE: 1500000 },
+  4: { STRING: "4개 일치 (50,000원)", PRICE: 50000 },
+  5: { STRING: "3개 일치 (5,000원)", PRICE: 5000 },
```
묶어서 관리하셨네요 ! 이 생각은 못했는데  객체로 관리해도 좋은 것 같아요

ssilver01 commented on src/Model/WinTicket.js:
```diff
@@ -0,0 +1,60 @@
+    for (let i = 1; i < userRank.length; i++) {
+      winMoney += userRank[i] * RANK[i].PRICE;
+    }
+    return Math.round((winMoney / purchase) * 1000) / 10;
```
1000 이부분도 rule에 있는   TICKET_PRICE로 바꿔도 좋을 것 같아요 !

ssilver01 reviewed (commented):
MVC패턴 적용한 점 너무 신기했네요. 사실 앱이나 서버에서만 적용해 보았지 js 코드로 된 건 처음 봐서 새로웠어요 ! 
각 Model에 함수 하나하나 신경쓴게 잘 보였고 한 함수가 하나의 기능만 하여 인상깊게 보았던 것 같아요 !
정말 많이 배워갑니다 👍

jiyeeeah committed:
[refactor] 당첨 통계를 계산하는 메서드 분리

- 당첨 통계를 바탕으로 등수를 return 하는 메서드 분리
- 8 - match를 왜 하는지 명시하기 위해 주석 추가

jiyeeeah committed:
[refactor] buyOneTicket -> buyTickets

- 굳이 한개의 티켓을 구매하는 함수를
  constructor에서 반복문으로 돌릴 필요가 없다고 판단
- buyTickets안에서 반복문 돌리도록 수정

jiyeeeah committed:
[refactor] 수익률 출력시, 천단위에 쉼표 찍히도록

jiyeeeah committed:
[feat] 보너스 번호 테스트 추가
